### ISSUE TITLE
Add get dependencies link to pubsec.yaml editor on add/change of a dependency

### ISSUE THREAD

**Issue by devoncarew** [flutter/flutter-intellij#2231]
_From @branflake2267 on July 8, 2017 20:24_

I keep adding dependencies and forgetting to click on get dependencies. I think this should fire automatically, or notify the user to download in the pubsec.yaml like it's done in the dart editor when they have been changed and haven't been downloaded. 

![screen shot 2017-07-08 at 1 22 15 pm](https://user-images.githubusercontent.com/1326504/27988807-c6dd0a2e-63e0-11e7-9eb9-eecceeb34706.png)

This may be redundant, so maybe another fresh notification to the links at top could be used...

## Steps to Reproduce
1. add a dependency
2. don't click on get dependencies
3. goto dart editor, and notice the download dependencies notification at the top of the editor

## Flutter Doctor

```
/Users/branflake2267/git/flutter/bin/flutter --no-color doctor
[✓] Flutter (on Mac OS X 10.12.5 16F73, locale en-US, channel master)
    • Flutter at /Users/branflake2267/git/flutter
    • Framework revision 4891506931 (21 hours ago), 2017-07-07 16:00:45 -0700
    • Engine revision 8ad1576f27
    • Tools Dart version 1.25.0-dev.4.0

[✓] Android toolchain - develop for Android devices (Android SDK 26.0.0)
    • Android SDK at /Users/branflake2267/Library/Android/sdk
    • Platform android-26, build-tools 26.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[-] iOS toolchain - develop for iOS devices (Xcode 8.3.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 8.3.3, Build version 8E3004b
    ✗ libimobiledevice and ideviceinstaller are not installed or require updating. To update, run:
        brew update
        brew uninstall --ignore-dependencies libimobiledevice
        brew install --HEAD libimobiledevice
        brew install ideviceinstaller
    • ios-deploy 1.9.1
    • CocoaPods version 1.2.1

[✓] Android Studio (version 2.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[✓] IntelliJ IDEA Community Edition (version 2017.1.5)
    • Flutter plugin version 15.1
    • Dart plugin version 171.4694.29

[✓] IntelliJ IDEA Ultimate Edition (version 2017.1.4)
    • Flutter plugin version 14.0
    • Dart plugin version 171.4694.29

[✓] Connected devices
    • iPhone 7 • 458BC4FE-8205-4369-8718-193B9FC88C8B • ios • iOS 10.3 (simulator)

Process finished with exit code 0
```

_Copied from original issue: flutter/flutter#11123_

**Comment by devoncarew** [comment]
_From @Hixie on July 8, 2017 21:21_

Are the dependencies needed for analysis and jump-to-definition and so on in the IDE?

**Comment by devoncarew** [comment]
_From @branflake2267 on July 8, 2017 21:36_

Nothing special, I'm adding the google_signin and firebase_auth dependencies. To say the least this is trivial. Android Studio gradle does things automatically so it feels laborious having to click on get packages or update packages.

**Comment by devoncarew** [comment]
_From @Hixie on May 8, 2018 23:23_

@devoncarew This sounds like an IntelliJ issue.

**Comment by pq** [comment]
I think this would be a great enhancement.  Since the pubspec editor lives in the Dart Plugin, it should likely get implemented there.

/cc @alexander-doroshko @jwren

**Comment by alexander-doroshko** [comment]
>Steps to Reproduce
>1. add a dependency
>2. don't click on get dependencies
>3. goto dart editor, and notice the download dependencies notification at the top of the editor

Right, for those who forget to click get Dependencies there's a reminder in the Dart file editor. You get the reminder exactly at the moment when you may need it. One click and you are done. 

A duplicating yellow bar in the pubspec.yaml file editor will be annoying for those who do not forget to do Get Dependencies. So far I'd prefer not adding it.

**Comment by pq** [comment]
What about a notification?  Like one of these?

![image](https://user-images.githubusercontent.com/67586/39878989-91e8fb68-542e-11e8-9031-a4310f354b92.png)

**Comment by alexander-doroshko** [comment]
At what moment do you want to get such notification?
What's wrong with the yellow bar in the Dart file editor? Why is it not enough?

**Comment by pq** [comment]
Good question.  I guess we might want a notification before they try and do anything that will fail if it hasn't been done.  The editor banner is good for that for editing and resolution.  Maybe we should guard running in case they try and run the app after a pubspec edit but never open a Dart file?

/cc @branflake2267

**Comment by branflake2267** [comment]
Just a note on the current behavior. After a dependency is added, I speed up the process by using save command, usually, using 'command + s' and then I fetch the dependencies. I've noticed that I have to save the file first before clicking on the menu shortcuts to get the dependencies. Whatever is polling for the file changes is slower than I like, which means I fetch and goto the main dart and have to fetch them again. It always seems a bit wonky when I fetch and have to do it again. 

Regarding a yellow bar, I don't know what you mean by that. So I'm assuming there is some other indicator or I don't have any memory of it. Maybe the yellow lightbulb? That would be handy next to the dependency if it weren't loaded yet, with a menu item to fetch or upgrade it maybe. 

That being said, I use the IDE to be lazy so I'm expecting the IDE to read the pubsec.yaml and just command the behavior. So really, I stated in the pubsec.yam I wanted the latest dependencies, so why couldn't it command the IDE to run the fetch and upgrade the dependencies each time the file is loaded and the project is loaded. If I don't want it to upgrade, I should specify the behavior in the versions of the dependency.  So at the end of the day, the manual fetch and upgrade logic feels verbose to me.

**Comment by alexander-doroshko** [comment]
> I speed up the process by using save command, usually, using 'command + s' and then I fetch the dependencies. I've noticed that I have to save the file first before clicking on the menu shortcuts to get the dependencies

That's strange. Get/Upgrade Dependencies action invoked from the IDE first saves all files and then runs pub command. In fact, there are very few cases when you might need to save files manually. If you have to save manually before clicking Get Dependencies it's a major bug that needs to be fixed.

**Comment by branflake2267** [comment]
I saw the yellow bar yesterday after I added a dependency, it shows up on the other dart libraries after I've added the library into pubsec.yaml and navigate to another dart library in the editor. So given I saw the yellow bar with notification to get the dependency or upgrade in another dart editor means I added the dependency to pubsec.yaml, and it wasn't downloaded by the time I navigated to the other editor. Maybe the library takes some time to download and cache and reflect that state change in the IDE, so I saw it before it was updated. Either case I couldn't tell anything was happening. Like gradle, where it has some notifications of actions happening in the bottom status bar.
